src/backend/CommandBufferStateTracker.cpp

```diff
@@ -90,96 +145,25 @@ namespace backend {
-            DAWN_RETURN_ERROR("Can't set the index buffer without a pipeline");
-        }
-
+    void CommandBufferStateTracker::SetIndexBuffer() {
         mAspects.set(VALIDATION_ASPECT_INDEX_BUFFER);
-        return {};
-    }
-
-    MaybeError CommandBufferStateTracker::SetVertexBuffer(uint32_t index) {
-        if (!HavePipeline()) {
-            DAWN_RETURN_ERROR("Can't set vertex buffers without a pipeline");
-        }
-
-        mInputsSet.set(index);
-        return {};
-    }
-
-    bool CommandBufferStateTracker::RecomputeHaveAspectBindGroups() {
-        if (mAspects[VALIDATION_ASPECT_BIND_GROUPS]) {
-            return true;
-        }
-        // Assumes we have a pipeline already
-        if (!mBindgroupsSet.all()) {
-            return false;
-        }
-        for (size_t i = 0; i < mBindgroups.size(); ++i) {
-            if (auto* bindgroup = mBindgroups[i]) {
-                // TODO(kainino@chromium.org): bind group compatibility
-                auto* pipelineBGL = mLastPipeline->GetLayout()->GetBindGroupLayout(i);
-                if (pipelineBGL && bindgroup->GetLayout() != pipelineBGL) {
-                    return false;
-                }
-            }
-        }
-        mAspects.set(VALIDATION_ASPECT_BIND_GROUPS);
-        return true;
-    }
-
-    bool CommandBufferStateTracker::RecomputeHaveAspectVertexBuffers() {
-        if (mAspects[VALIDATION_ASPECT_VERTEX_BUFFERS]) {
-            return true;
-        }
-        // Assumes we have a pipeline already
-        auto requiredInputs = mLastRenderPipeline->GetInputState()->GetInputsSetMask();
-        if ((mInputsSet & requiredInputs) == requiredInputs) {
-            mAspects.set(VALIDATION_ASPECT_VERTEX_BUFFERS);
-            return true;
-        }
-        return false;
-    }
-
-    bool CommandBufferStateTracker::HavePipeline() const {
-        return mAspects[VALIDATION_ASPECT_PIPELINE];
     }
 
-    MaybeError CommandBufferStateTracker::RevalidateCanDraw() {
-        if (!mAspects[VALIDATION_ASPECT_PIPELINE]) {
-            DAWN_RETURN_ERROR("No active render pipeline");
-        }
-        // Compute the lazily computed mAspects
-        if (!RecomputeHaveAspectBindGroups()) {
-            DAWN_RETURN_ERROR("Bind group state not valid");
+    void CommandBufferStateTracker::SetVertexBuffer(uint32_t start, uint32_t count) {
+        for (uint32_t i = 0; i < count; ++i) {
+            mInputsSet.set(start + i);
         }
-        if (!RecomputeHaveAspectVertexBuffers()) {
-            DAWN_RETURN_ERROR("Some vertex buffers are not set");
-        }
-        return {};
     }
 
     void CommandBufferStateTracker::SetPipelineCommon(PipelineBase* pipeline) {
-        PipelineLayoutBase* layout = pipeline->GetLayout();
+        mLastPipelineLayout = pipeline->GetLayout();
 
         mAspects.set(VALIDATION_ASPECT_PIPELINE);
-
         mAspects.reset(VALIDATION_ASPECT_BIND_GROUPS);
         mAspects.reset(VALIDATION_ASPECT_VERTEX_BUFFERS);
```

Comment:
Probably an existing issue, but should this also reset `VALIDATION_ASPECT_INDEX_BUFFER`?